src/Microsoft.OData.Edm/Csdl/Parsing/SchemaJsonParser.cs

```diff
@@ -0,0 +1,1564 @@
+                }
+            });
+
+            CsdlComplexType csdlComplexType = new CsdlComplexType(name,
+                baseType,
+                isAbstract == null ? false : isAbstract.Value, // default is false
+                isOpen == null ? false : isOpen.Value, // default is false
+                members.OfType<CsdlProperty>(),
+                members.OfType<CsdlNavigationProperty>(),
+                context.Location());
+
+            members.OfType<CsdlAnnotation>().ForEach(a => csdlComplexType.AddAnnotation(a));
+            return csdlComplexType;
+        }
+
+        /// <summary>
+        /// Parse the <see cref="JsonElement"/> to <see cref="CsdlKey"/>.
+        /// </summary>
+        /// <param name="name">The name.</param>
+        /// <param name="jsonValue">The json value to parse from.</param>
+        /// <param name="jsonPath">The json path of this value.</param>
+        /// <returns>the parsed <see cref="CsdlKey"/>.</returns>
+        internal static CsdlKey ParseCsdlKey(string name, JsonElement keyArray, JsonParserContext context)
+        {
+            // The value of $Key is an array with one item per key property.
+            Debug.Assert(keyArray.ValueKind == JsonValueKind.Array);
+            Debug.Assert(name == "$Key", "The name should be $Key.");
+
+            IList<CsdlPropertyReference> properties = keyArray.ParseAsArray(context, (v, p) =>
+            {
+                if (v.ValueKind == JsonValueKind.Object)
+                {
+                    // TODO: ODL doesn't support the key referencing a property of a complex type as below.
+                    // "$Key": [
+                    //    {
+                    //      "EntityInfoID": "Info/ID"
+                    //    }
+                    //  ]
+                    p.ReportError(EdmErrorCode.JsonInvalidKeyValue, "It doesn't support the key object.");
+                }
+
+                string propertyName = v.ParseAsString(context);
+                return new CsdlPropertyReference(propertyName, context.Location());
+            });
+
+            return new CsdlKey(properties, context.Location());
+        }
+
+        /// <summary>
+        /// Parse the <see cref="JsonElement"/> to <see cref="CsdlProperty"/>.
+        /// </summary>
+        /// <param name="name">The name.</param>
+        /// <param name="propertyObject">The json value to parse from.</param>
+        /// <param name="jsonPath">The json path of this value.</param>
+        /// <returns>the parsed <see cref="CsdlProperty"/>.</returns>
+        internal static CsdlProperty ParseCsdlProperty(string name, JsonElement propertyObject, JsonParserContext context)
+        {
+            // Structural properties are represented as members of the object representing a structured type.
+            // The member name is the property name, the member value is an object.
+            Debug.Assert(propertyObject.ValueKind == JsonValueKind.Object);

```

Comment:
> Debug.Assert(propertyObject.ValueKind == JsonValueKind.Object); [](start = 11, length = 64)

should this be an assert or an error?  Are you asserting that we've already validated that it's an object?